modules/core/04-channel/v2/keeper/msg_server.go

```diff
@@ -111,26 +104,45 @@ func (k *Keeper) RecvPacket(ctx context.Context, msg *types.MsgRecvPacket) (*typ
 			if len(msg.Packet.Payloads) > 1 {
 				return nil, errorsmod.Wrapf(types.ErrInvalidPacket, "packet with multiple payloads cannot have async acknowledgement")
 			}
+			// break out of the loop if the acknowledgement is async
+			// to prevent the addition of nil to the acknowledgement
+			break
 		}
 
 		// append app acknowledgement to the overall acknowledgement
 		ack.AppAcknowledgements = append(ack.AppAcknowledgements, res.Acknowledgement)
-	}
 
-	if len(ack.AppAcknowledgements) != len(msg.Packet.Payloads) {
-		return nil, errorsmod.Wrapf(types.ErrInvalidAcknowledgement, "length of app acknowledgement %d does not match length of app payload %d", len(ack.AppAcknowledgements), len(msg.Packet.Payloads))
+		// if any of the application callbacks return a failure, the entire packet receive is considered a failure
+		// break the acknowledgement loop but keep a single non-nil acknowledgement in the acknowledgement slice
+		// so that the error acknowledgment is not equivalent to the empty acknowledgement
+		if res.Status == types.PacketStatus_Failure {
```

Comment:
I moved this down so that specific app acknowledgement is filled in even in failure with single packet payload